### DISCUSSION TITLE
Download and installation automation improvements

### DISCUSSION THREAD

**Discussion by ossfellow** [FiloSottile/age#292] [General]
Hello folks,
As part of the v1 release, could you please improve how releases are packaged and named?
This will improve how Age installation is added to scripts, Dockerfiles, etc.

Some observations that make it difficult as-is are:

- Contents of the tar file are different in different releases, which makes automation impossible, or unnecessarily cumbersome (extra coding and checking about the presence of all components).
- The tar file contains a container directory (age/), which is unnecessary and makes scripting cumbersome.
- Version number is both in the download link, and the package name. The latter seems really unnecessary.
 
So, it would be great if scripting and automation friendliness is incorporated early on, as right now it is not straightforward.

Thank you.

**Comment by FiloSottile** [comment]
Hello! Thank you for the feedback. Could you elaborate or provide an example on what you mean that the contents are different?

As for the container directory, it's an explicit choice, because if you extract the tarball you don't want a random LICENSE file laying around. If by package name you mean the tar or zip filename, the version number is also an explicit choice, because that way different version can be told apart. As far as I know that's consistent with most releases on GitHub.

**Reply by ossfellow** [reply]
Please look into the contents of rc2 and rc3 for Linux. rc3 only has age binary.
If one use --strip to remove the top container in a script, or manually, the LICENSE file ends up in the destination anyways; that’s partly why top container is not part of convention.

**Reply by FiloSottile** [reply]
`age-v1.0.0-rc.3-linux-amd64.tar.gz` looks fine, which file specifically?

```
$ tar tf age-v1.0.0-rc.2-linux-amd64.tar.gz
age/
age/age
age/age-keygen
age/LICENSE
$ tar tf age-v1.0.0-rc.3-linux-amd64.tar.gz
age/
age/age
age/age-keygen
age/LICENSE
```

> If one use --strip to remove the top container in a script, or manually, the LICENSE file ends up in the destination anyways; that’s partly why top container is not part of convention.

I don't follow, the availability of `--strip-components` is an argument for putting in a container: if automation doesn't want it, they can just use `--strip-components` in a script, while a lack of container can be unexpected using default settings and annoying to cleanup.

**Comment by alerque** [comment]
As a distro packager (and former maintainer of the `age` package on Arch) I'd like to register an objection to some of the points.

1. Of course consistency between releases is a good thing.
2. The container directory in the tar *should* be there, nothing should extract to a bare directory. Doing so make many situations more troublesome than they need to be, such as building or comparing multiple versions. In fact the current scheme is a problem because the embedded directory doesn't have a version. It should extract to `age-vX.Y.Z` as the top level directory.
3. The downloaded tarball file *should* have the version number in it.

If you want to cater to people that just want to `wget` a binary straight off GitHub then also attach the raw binaries to the release, perhaps without version info in the name, and but that's not what I expect when I download a tarball (even a of binary releases).

**Comment by ossfellow** [comment]
Thank you gentlemen for your feedback. 
I use the age package in a dockerfile; so, it’s not an interactive use case. For reasons unknown to me, given that the tar extraction command hasn’t changed there, the build using rc3 package had only extracted the age binary; rc2 build was fine; so, I took that as a sign of inconsistency.
Funny thing is that a new build using rc3 is fine too; so, I can’t reproduce the issue. Nonetheless, whatever the packaging preferences of the maintainers, as a user, as long as they remain consistent, then I can code accordingly.
I’m good here, thank you.

**Comment by FiloSottile** [comment]
I added redirect links for specific and latest releases. They point to the same artifacts as before, but you can save them to a fixed name, at which point the version won't be in the way.

https://dl.filippo.io/age/latest?for=linux/arm64
https://dl.filippo.io/age/v1.0.0-rc.1?for=darwin/arm64
